rfcs/2021-08-29-8381-vrl-iteration-support.md

```diff
@@ -0,0 +1,756 @@
+in regular blocks as well.
+
+While there is _some_ inconsistency here, the inconsistency is well-isolated, it
+prevents us from having to introduce a breaking change right now, and it also
+still leaves open the option of adding lexical scoping before a 1.0 release of
+VRL.
+
+#### Returning Two-Element Array
+
+We require the `map` function closure to return a two-element `array` type.
+Without this requirement, mapping would work as follows:
+
+```coffee
+map(.) |key, _| {
+  key = upcase(key)
+}
+```
+
+That is, `key` would be a "special variable" inside the closure, which modifies
+the actual key of the record within the object.
+
+This doesn't fit existing patterns in VRL. It looks as if there's a _dangling_
+variable `key` at the end that remains unused, but because we special-cased this
+situation, it would instead magically update the actual key in the object after
+the closure runs to completion.
+
+This can become more difficult to reason about if/when we introduce control-flow
+statements such as `break`, as you could have set `key` before calling `break`,
+which would then either still mutate the actual key, or not, depending on how we
+implement `break`, but either way, the program itself becomes less readable, and
+operators have to read the language documentation to understand the semantic
+differences between how code behaves _inside_ a function-closure and _outside_.
+
+Instead, the `map` function-closure is required to return a two-element array of
+`[key, value]`, which the function machinery then uses to update the actual
+values of the object record, e.g.:
+
+```coffee
+map(.) |key, value| {
+  key = upcase(key)
+
+  # The array return-value clearly defines the eventual key and value values.
+  [key, value]
+}
+```
+
+##### Tuple Type
+
+Alternatively, we could introduce a new `tuple` type to define the return-type
+of the closure:
+
+```coffee
+map(.) |key, value| {
+  key = upcase(key)
+
+  (key, value)
+}
+```
+
+They would semantically be the same, given that VRL supports mixing value types
+in arrays, which is usually the difference between a tuple and an array, that,
+and the fact that a tuple is immutable, but an array doesn't have to be.
+
+Semantically, using a tuple makes more sense, but it does add an extra `Value`
+type, and it does mean we have to convert that tuple to a type supported by
+JSON, which will likely be an array, so the external JSON representation of
+tuples remains the same, regardless of whether we use arrays or tuples inside
+VRL itself.
+
+Since there isn't a clear benefit at this moment to using a tuple over
+a two-element array, the choice is made to forgo adding the tuple type at this
+moment.
+
+#### Parser Ambiguity
+
+TODO, figure out how to handle parser ambiguity.
+
+That is, these are identical:
+
+```coffee
+# merge the result of `fn(a)` with `b` and `c`
+fn(a) | b | c
+```
+
+```coffee
+# call `fn(a)` with closure `|b| c`
+fn(a) |b| c
+```
+
```

Comment:
I think we should require the blocks to be surrounded by `{...}`. I believe Ruby requires you to put the `{` before the parameter list which should fix this.

```ruby
fn(a) { |b| c }
```

Reply:
I'm irrationally against that syntax, it "doesn't feel right to me". I agree though that it would help with the ambiguity, so it's worth considering.

I do very much like the non-block closure syntax for simple iterations, but I wouldn't be against not allowing that, at least not in a V1 for simplicity reasons.

Reply:
I came around to:

- not minding us losing short-hand notation of closures, it's only two less characters (and, as @pablosichert mentioned, it can result in more cognitive complexity when it comes to operator precedence).
- not being a fan of the `-> { ... }` syntax

The new syntax is now as you proposed:

```coffee
fn(a) { |b| c }
```